hydrator-common/src/main/java/co/cask/hydrator/common/TimeParser.java

```diff
@@ -0,0 +1,158 @@
+ * the License.
+ */
+
+package co.cask.hydrator.common;
+
+import com.google.common.base.Preconditions;
+import com.google.common.base.Strings;
+
+import java.util.concurrent.TimeUnit;
+
+/**
+ * Parses time expressions and does simple time math.
+ */
+public class TimeParser {
+  private final long runtime;
+
+  public TimeParser(long runtime) {
+    this.runtime = runtime;
+  }
+
+  /**
+   * Parses time strings into timestamps, with support for some basic math.
+   * Math syntax includes addition and subtraction in seconds, minutes, hours, and days.
+   * Periods of time are specified as some number followed by the time unit,
+   * where 's' is seconds', 'm' is minutes, 'h' is hours, and 'd' is days.
+   * The "runtime" keyword translates to some given runtime, and can be strung together with
+   * various additions and subtractions.
+   * For example, "runtime-5s" is 5 seconds before runtime, "runtime-1d" is one day before runtime,
+   * and "runtime-1d+4h" is 20 hours before runtime.
+   *
+   * @param str the time string to parse
+   * @return milliseconds equivalent of the time string
+   */
+  public long parseRuntime(String str) {
+    MathNode mathNode = parseExpression(str);
+    return mathNode.evaluate();
+  }
+
+  /**
+   * Parses a duration String to its long value.
+   * Frequency string consists of a number followed by an unit, with 's' for seconds, 'm' for minutes, 'h' for hours
+   * and 'd' for days. For example, an input of '5m' means 5 minutes which will be parsed to 300000 milliseconds.
+   *
+   * @param durationStr the duration string (ex: 5m, 5h etc).
+   * @return milliseconds equivalent of the duration string
+   */
+  public static long parseDuration(String durationStr) {
+    Preconditions.checkArgument(!Strings.isNullOrEmpty(durationStr));
+    durationStr = durationStr.trim().toLowerCase();
+
+    String value = durationStr.substring(0, durationStr.length() - 1);
+    long parsedValue = Long.parseLong(value);
+
+    char lastChar = durationStr.charAt(durationStr.length() - 1);
+    switch (lastChar) {
+      case 's':
+        return TimeUnit.SECONDS.toMillis(parsedValue);
+      case 'm':
+        return TimeUnit.MINUTES.toMillis(parsedValue);
+      case 'h':
+        return TimeUnit.HOURS.toMillis(parsedValue);
+      case 'd':
+        return TimeUnit.DAYS.toMillis(parsedValue);
+    }
+    throw new IllegalArgumentException(String.format("Time unit not supported: %s", lastChar));
+  }
+
+  // parse an expression into an evaluation tree
+  private MathNode parseExpression(String expression) {
+    expression = expression.trim();
+    // we only support + and - right now so just evaluate from left to right
+    int idx = Math.max(expression.lastIndexOf('+'), expression.lastIndexOf('-'));
+
+    // if no operator, this should be a duration string
+    if (idx < 0) {
+      return "runtime".equals(expression) ? new ValueNode(runtime) : new ValueNode(parseDuration(expression));
```

Comment:
they are different.  Here it is part of an expression, whereas Macros defines a macro type called 'runtime'.